evm/vm/forks/sharding/guess_head_state_manager.py

```diff
@@ -0,0 +1,245 @@
+import asyncio
+from collections import (
+    defaultdict,
+)
+import logging
+import time
+
+from evm.vm.forks.sharding.constants import (
+    GENESIS_COLLATION_HASH,
+)
+from evm.vm.forks.sharding.shard_tracker import (
+    NoCandidateHead,
+)
+
+
+NUM_RESERVED_BLOCKS = 5
+SIMULATED_COLLATION_DOWNLOADING_TIME = 2
+SIMULATED_COLLATION_VERIFICATION_TIME = 0.01
+
+
+logger = logging.getLogger("evm.chain.sharding.guess_head_state_manager")
+
+
+async def download_collation(collation_hash):
+    # TODO: need to implemented after p2p part is done
+    logger.debug("Start downloading collation %s", collation_hash)
+    await asyncio.sleep(SIMULATED_COLLATION_DOWNLOADING_TIME)
+    logger.debug("Finished downloading collation %s", collation_hash)
+    collation = collation_hash
+    return collation
+
+
+async def verify_collation(collation, collation_hash):
+    # TODO: to be implemented
+    logger.debug("Verifying collation %s", collation_hash)
+    await asyncio.sleep(SIMULATED_COLLATION_VERIFICATION_TIME)
+    # data = get_from_p2p_network(collation_hash.data_root)
+    # chunks = DATA_SIZE // 32
+    # mtree = [0] * chunks + [data[chunks*32: chunks*32+32] for i in range(chunks)]
+    # for i in range(chunks-1, 0, -1):
+    #     mtree[i] = sha3(mtree[i*2] + mtree[i*2+1])
+    # assert mtree[i] == collation.data_root
+    return True
+
+
+async def fetch_and_verify_collation(collation_hash):
+    """Fetch the collation body and verify the collation
+
+    :return: returns the collation's validity
+    """
+    collation = await download_collation(collation_hash)
+    return await verify_collation(collation, collation_hash)
+
+
+def clean_done_task(tasks):
+    return [task for task in tasks if not task.done()]
+
+
+class GuessHeadStateManager:
```

Comment:
> IMO:
This class name is too long.
It's unclear what's the "state" for.
ghs_manager looks like a ghost manager.

I agree. Will change them.

> It's inefficient if you are setting each collator has to maintain SHARD_COUNT GuessHeadStateManagers. Does this process be initialized for each known period that the collator applied lookahead and knows it's the eligible collator?

This class is only initialized once, and able to be reused in the future `guess_head`s. My thought about the inefficiency issue is: we need one `log_handler` for each shard, because `get_new_logs` is called for each shard at different moment. And every `shard_tracker` needs one `log_handler`. Even though we move `guess_head` to `shard_tracker` and collator related logics to class `Collator`, we still need `SHARD_COUNT` `shard_tracker`s for one collator. Do you have any solution in mind already? You're right It might be able to be further optimized.

> If yes, it's like a worker to do its job - windback? If so, how about WindbackWorker...?

I'm fine with this name. It's better than the original one.

Reply:
If `LOOKAHEAD_LENGTH := 4`, let's say the collator disconnect after one more period, the maximum concurrently existing `log_handler` is 5, I think it's tolerable if the separated-period-workers could simplify many sharing memory problems.

Reply:
Yes you're right. If you're concern is `SHARD_COUNT` `log_handler`s, actually we only store the recent `256` blockhashes. So roughly there will be 256*32 bytes*100 = 819200 bytes <= 1 MB memory consumption, if a collator synchronizes all shards.
IMO it seems for now it is not a problem? I can also mark a TODO to indicate this issue so that we can do further optimization later?